test/extensions/filters/http/jwt_authn/group_verifier_test.cc

```diff
@@ -0,0 +1,316 @@
+  auto headers = Http::TestHeaderMapImpl{
+      {":path", "/requires-all?jwt_a=xxx&jwt_b=xxx"},
+      {"example-auth-userinfo", ""},
+      {"other-auth-userinfo", ""},
+  };
+  context_ = Verifier::createContext(headers, &mock_cb_);
+  verifier_->verify(context_);
+  callbacks[0](Status::JwtBadFormat);
+  // can keep invoking callback
+  callbacks[1](Status::Ok);
+  callbacks[0](Status::Ok);
+  callbacks[1](Status::Ok);
+
+  EXPECT_FALSE(headers.has("example-auth-userinfo"));
+  EXPECT_FALSE(headers.has("other-auth-userinfo"));
+}
+
+// test requires all with second token returning missing jwt
+TEST_F(GroupVerifierTest, TestRequiresAllMissing) {
+  MessageUtil::loadFromYaml(RequiresAllConfig, proto_config_);
+  std::vector<AuthenticatorCallback> callbacks;
+  for (int i = 0; i < 2; ++i) {
+    auto mock_auth = std::make_unique<MockAuthenticator>();
+    EXPECT_CALL(*mock_auth.get(), doVerify(_, _, _))
+        .WillRepeatedly(Invoke(
+            [&](Http::HeaderMap&, std::vector<JwtLocationConstPtr>*,
+                AuthenticatorCallback callback) { callbacks.push_back(std::move(callback)); }));
+    mock_auths_.push_back(std::move(mock_auth));
+  }
+  createVerifier();
+
+  EXPECT_CALL(mock_cb_, onComplete(_)).WillOnce(Invoke([](const Status& status) {
+    ASSERT_EQ(status, Status::JwtMissed);
+  }));
+  auto headers = Http::TestHeaderMapImpl{
+      {":path", "/requires-all?jwt_a=xxx&jwt_b=xxx"},
+      {"example-auth-userinfo", ""},
+      {"other-auth-userinfo", ""},
+  };
+  context_ = Verifier::createContext(headers, &mock_cb_);
+  verifier_->verify(context_);
+  callbacks[0](Status::Ok);
+  callbacks[1](Status::JwtMissed);
+  // can keep invoking callback
+  callbacks[0](Status::Ok);
+  callbacks[1](Status::Ok);
+  EXPECT_FALSE(headers.has("example-auth-userinfo"));
+  EXPECT_FALSE(headers.has("other-auth-userinfo"));
+}
+
+TEST_F(GroupVerifierTest, TestRequiresAllWrongLocations) {
+  MessageUtil::loadFromYaml(RequiresAllConfig, proto_config_);
+  auto mock_auth = std::make_unique<MockAuthenticator>();
+  EXPECT_CALL(*mock_auth.get(), doVerify(_, _, _))
+      .WillRepeatedly(
+          Invoke([](Http::HeaderMap&, std::vector<JwtLocationConstPtr>*,
+                    AuthenticatorCallback callback) { callback(Status::JwtUnknownIssuer); }));
+  mock_auths_.push_back(std::move(mock_auth));
+  createVerifier();
+
+  EXPECT_CALL(mock_cb_, onComplete(_)).WillOnce(Invoke([](const Status& status) {
+    ASSERT_EQ(status, Status::JwtUnknownIssuer);
+  }));
+  auto headers = Http::TestHeaderMapImpl{
+      {":path",
+       "/requires-all?jwt_a=" + std::string(OtherGoodToken) + "&jwt_b=" + std::string(GoodToken)},
+      {"example-auth-userinfo", ""},
+      {"other-auth-userinfo", ""},
+  };
+  context_ = Verifier::createContext(headers, &mock_cb_);
+  verifier_->verify(context_);
+  EXPECT_FALSE(headers.has("example-auth-userinfo"));
+  EXPECT_TRUE(headers.has("other-auth-userinfo"));
+}
+
+TEST_F(GroupVerifierTest, TestRequiresAny) {
```

Comment:
I like you to split "require_any" test case into smaller ones, each cover one specific case
Also add some nested group case, such as "all" nested with "any", and "any" nested with "all".